<div style="border:solid Chocolate 2px; padding: 40px">

<b>Денис, приветствую!</b>

Буду рада с тобой познакомиться. Меня зовут Евгения Маракова и я буду ревьюером твоего проекта. Если ты не против, то предлагаю общаться на "ты":) Если удобнее на "вы", то нет проблем, только скажи об этом.

В процессе работы ты встретишь комментарии <font color='green'>зеленого</font>, <font color='gold'>желтого</font> и <font color='red'>красного</font> цветов. Что они значат, смотри ниже:
<br/>

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Отлично!✅:</b> Решение на этом шаге является полностью правильным.

</div>

<br/>

<div class="alert alert-warning">
    <h2> Комментарий ревьюера <a class="tocSkip"> </h2>
    
<b>Можно лучше⚠️:</b> Решение на этом шаге станет еще лучше, если внести небольшие коррективы.
</div>

<br/>
<div class="alert alert-block alert-danger">
<h2> Комментарий ревьюера <a class="tocSkip"></h2>

<b>Необходимо исправить❌:</b>
Решение на этом шаге требует существенной переработки и внесения правок. Напоминаю, что проект не может быть принят с первого раза, если ревью содержит комментарии, рекомендующие доработать шаги.

</div>
    
Увидев мой комментарий, не удаляй его, он будет очень полезен в случае повторной проверки работы😊
    
Свой комментарий помечай так, как показано ниже, тогда нам будет легче вести диалог, а мне выполнить повторную проверку проекта.
    <div class="alert alert-info"> <b>Комментарий студента: </b> </div>


# Исследование надежности заемщиков

Во второй части проекта вы выполните шаги 3 и 4. Их вручную проверит ревьюер.
Чтобы вам не пришлось писать код заново для шагов 1 и 2, мы добавили авторские решения в ячейки с кодом.


## Откройте таблицу и изучите общую информацию о данных


**Задание 1. Импортируйте библиотеку pandas. Считайте данные из csv-файла в датафрейм и сохраните в переменную `data`. Путь к файлу:**

`/datasets/data.csv`


In [2]:
import pandas as pd

try:
    data = pd.read_csv("/datasets/data.csv")
except:
    data = pd.read_csv("https://code.s3.yandex.net/datasets/data.csv")

**Задание 2. Выведите первые 20 строчек датафрейма `data` на экран.**


In [4]:
data.head(20)

children  days_employed  dob_years            education  education_id  \
0          1   -8437.673028         42               высшее             0   
1          1   -4024.803754         36              среднее             1   
2          0   -5623.422610         33              Среднее             1   
3          3   -4124.747207         32              среднее             1   
4          0  340266.072047         53              среднее             1   
5          0    -926.185831         27               высшее             0   
6          0   -2879.202052         43               высшее             0   
7          0    -152.779569         50              СРЕДНЕЕ             1   
8          2   -6929.865299         35               ВЫСШЕЕ             0   
9          0   -2188.756445         41              среднее             1   
10         2   -4171.483647         36               высшее             0   
11         0    -792.701887         40              среднее             1   
12         0            NaN         65              среднее             1   
13         0   -1846.641941         54  неоконченное высшее             2   
14         0   -1844.956182         56               высшее             0   
15         1    -972.364419         26              среднее             1   
16         0   -1719.934226         35              среднее             1   
17         0   -2369.999720         33               высшее             0   
18         0  400281.136913         53              среднее             1   
19         0  -10038.818549         48              СРЕДНЕЕ             1   

       family_status  family_status_id gender income_type  debt  \
0    женат / замужем                 0      F   сотрудник     0   
1    женат / замужем                 0      F   сотрудник     0   
2    женат / замужем                 0      M   сотрудник     0   
3    женат / замужем                 0      M   сотрудник     0   
4   гражданский брак                 1      F   пенсионер     0   
5   гражданский брак                 1      M   компаньон     0   
6    женат / замужем                 0      F   компаньон     0   
7    женат / замужем                 0      M   сотрудник     0   
8   гражданский брак                 1      F   сотрудник     0   
9    женат / замужем                 0      M   сотрудник     0   
10   женат / замужем                 0      M   компаньон     0   
11   женат / замужем                 0      F   сотрудник     0   
12  гражданский брак                 1      M   пенсионер     0   
13   женат / замужем                 0      F   сотрудник     0   
14  гражданский брак                 1      F   компаньон     1   
15   женат / замужем                 0      F   сотрудник     0   
16   женат / замужем                 0      F   сотрудник     0   
17  гражданский брак                 1      M   сотрудник     0   
18    вдовец / вдова                 2      F   пенсионер     0   
19         в разводе                 3      F   сотрудник     0   

     total_income                                 purpose  
0   253875.639453                           покупка жилья  
1   112080.014102                 приобретение автомобиля  
2   145885.952297                           покупка жилья  
3   267628.550329              дополнительное образование  
4   158616.077870                         сыграть свадьбу  
5   255763.565419                           покупка жилья  
6   240525.971920                       операции с жильем  
7   135823.934197                             образование  
8    95856.832424                   на проведение свадьбы  
9   144425.938277                 покупка жилья для семьи  
10  113943.491460                    покупка недвижимости  
11   77069.234271       покупка коммерческой недвижимости  
12            NaN                         сыграть свадьбу  
13  130458.228857                 приобретение автомобиля  
14  165127.911772              покупка жилой недвижимости  
15  116820.904450 

**Задание 3. Выведите основную информацию о датафрейме с помощью метода `info()`.**


In [37]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 21525 entries, 0 to 21524
Data columns (total 12 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   children          21525 non-null  int64  
 1   days_employed     19351 non-null  float64
 2   dob_years         21525 non-null  int64  
 3   education         21525 non-null  object 
 4   education_id      21525 non-null  int64  
 5   family_status     21525 non-null  object 
 6   family_status_id  21525 non-null  int64  
 7   gender            21525 non-null  object 
 8   income_type       21525 non-null  object 
 9   debt              21525 non-null  int64  
 10  total_income      19351 non-null  float64
 11  purpose           21525 non-null  object 
dtypes: float64(2), int64(5), object(5)
memory usage: 2.0+ MB


## Предобработка данных


### Удаление пропусков


**Задание 4. Выведите количество пропущенных значений для каждого столбца. Используйте комбинацию двух методов.**


In [38]:
data.isna().sum()

children               0
days_employed       2174
dob_years              0
education              0
education_id           0
family_status          0
family_status_id       0
gender                 0
income_type            0
debt                   0
total_income        2174
purpose                0
dtype: int64

**Задание 5. В двух столбцах есть пропущенные значения. Один из них — `days_employed`. Пропуски в этом столбце вы обработаете на следующем этапе. Другой столбец с пропущенными значениями — `total_income` — хранит данные о доходах. На сумму дохода сильнее всего влияет тип занятости, поэтому заполнить пропуски в этом столбце нужно медианным значением по каждому типу из столбца `income_type`. Например, у человека с типом занятости `сотрудник` пропуск в столбце `total_income` должен быть заполнен медианным доходом среди всех записей с тем же типом.**


In [39]:
for t in data["income_type"].unique():
    data.loc[
        (data["income_type"] == t) & (
            data["total_income"].isna()), "total_income"
    ] = data.loc[(data["income_type"] == t), "total_income"].median()

### Обработка аномальных значений


**Задание 6. В данных могут встречаться артефакты (аномалии) — значения, которые не отражают действительность и появились по какой-то ошибке. таким артефактом будет отрицательное количество дней трудового стажа в столбце `days_employed`. Для реальных данных это нормально. Обработайте значения в этом столбце: замените все отрицательные значения положительными с помощью метода `abs()`.**


In [40]:
data["days_employed"] = data["days_employed"].abs()

**Задание 7. Для каждого типа занятости выведите медианное значение трудового стажа `days_employed` в днях.**


In [41]:
data.groupby("income_type")["days_employed"].agg("median")

income_type
безработный        366413.652744
в декрете            3296.759962
госслужащий          2689.368353
компаньон            1547.382223
пенсионер          365213.306266
предприниматель       520.848083
сотрудник            1574.202821
студент               578.751554
Name: days_employed, dtype: float64

У двух типов (безработные и пенсионеры) получатся аномально большие значения. Исправить такие значения сложно, поэтому оставьте их как есть. Тем более этот столбец не понадобится вам для исследования.


**Задание 8. Выведите перечень уникальных значений столбца `children`.**


In [42]:
data["children"].unique()

array([ 1,  0,  3,  2, -1,  4, 20,  5], dtype=int64)

**Задание 9. В столбце `children` есть два аномальных значения. Удалите строки, в которых встречаются такие аномальные значения из датафрейма `data`.**


In [43]:
data = data[(data["children"] != -1) & (data["children"] != 20)]

**Задание 10. Ещё раз выведите перечень уникальных значений столбца `children`, чтобы убедиться, что артефакты удалены.**


In [44]:
data["children"].unique()

array([1, 0, 3, 2, 4, 5], dtype=int64)

### Удаление пропусков (продолжение)


**Задание 11. Заполните пропуски в столбце `days_employed` медианными значениями по каждого типа занятости `income_type`.**


In [45]:
for t in data["income_type"].unique():
    data.loc[
        (data["income_type"] == t) & (
            data["days_employed"].isna()), "days_employed"
    ] = data.loc[(data["income_type"] == t), "days_employed"].median()

**Задание 12. Убедитесь, что все пропуски заполнены. Проверьте себя и ещё раз выведите количество пропущенных значений для каждого столбца с помощью двух методов.**


In [46]:
data.isna().sum()

children            0
days_employed       0
dob_years           0
education           0
education_id        0
family_status       0
family_status_id    0
gender              0
income_type         0
debt                0
total_income        0
purpose             0
dtype: int64

### Изменение типов данных


**Задание 13. Замените вещественный тип данных в столбце `total_income` на целочисленный с помощью метода `astype()`.**


In [47]:
data["total_income"] = data["total_income"].astype(int)

### Обработка дубликатов


**Задание 14. Обработайте неявные дубликаты в столбце `education`. В этом столбце есть одни и те же значения, но записанные по-разному: с использованием заглавных и строчных букв. Приведите их к нижнему регистру. Проверьте остальные столбцы.**


In [48]:
data["education"] = data["education"].str.lower()

**Задание 15. Выведите на экран количество строк-дубликатов в данных. Если такие строки присутствуют, удалите их.**


In [49]:
data.duplicated().sum()

71

In [50]:
data = data.drop_duplicates()

### Категоризация данных


**Задание 16. На основании диапазонов, указанных ниже, создайте в датафрейме `data` столбец `total_income_category` с категориями:**

- 0–30000 — `'E'`;
- 30001–50000 — `'D'`;
- 50001–200000 — `'C'`;
- 200001–1000000 — `'B'`;
- 1000001 и выше — `'A'`.

**Например, кредитополучателю с доходом 25000 нужно назначить категорию `'E'`, а клиенту, получающему 235000, — `'B'`. Используйте собственную функцию с именем `categorize_income()` и метод `apply()`.**


In [51]:
def categorize_income(income):
    try:
        if 0 <= income <= 30000:
            return "E"
        elif 30001 <= income <= 50000:
            return "D"
        elif 50001 <= income <= 200000:
            return "C"
        elif 200001 <= income <= 1000000:
            return "B"
        elif income >= 1000001:
            return "A"
    except:
        pass

In [52]:
data["total_income_category"] = data["total_income"].apply(categorize_income)

**Задание 17. Выведите на экран перечень уникальных целей взятия кредита из столбца `purpose`.**


In [53]:
data["purpose"].unique()
data

children  days_employed  dob_years education  education_id  \
0             1    8437.673028         42    высшее             0   
1             1    4024.803754         36   среднее             1   
2             0    5623.422610         33   среднее             1   
3             3    4124.747207         32   среднее             1   
4             0  340266.072047         53   среднее             1   
...         ...            ...        ...       ...           ...   
21520         1    4529.316663         43   среднее             1   
21521         0  343937.404131         67   среднее             1   
21522         1    2113.346888         38   среднее             1   
21523         3    3112.481705         38   среднее             1   
21524         2    1984.507589         40   среднее             1   

          family_status  family_status_id gender income_type  debt  \
0       женат / замужем                 0      F   сотрудник     0   
1       женат / замужем                 0      F   сотрудник     0   
2       женат / замужем                 0      M   сотрудник     0   
3       женат / замужем                 0      M   сотрудник     0   
4      гражданский брак                 1      F   пенсионер     0   
...                 ...               ...    ...         ...   ...   
21520  гражданский брак                 1      F   компаньон     0   
21521   женат / замужем                 0      F   пенсионер     0   
21522  гражданский брак                 1      M   сотрудник     1   
21523   женат / замужем                 0      M   сотрудник     1   
21524   женат / замужем                 0      F   сотрудник     0   

       total_income                       purpose total_income_category  
0            253875                 покупка жилья                     B  
1            112080       приобретение автомобиля                     C  
2            145885                 покупка жилья                     C  
3            267628    дополнительное образование                     B  
4            158616               сыграть свадьбу                     C  
...             ...                           ...                   ...  
21520        224791             операции с жильем                     B  
21521        155999          сделка с автомобилем                     C  
21522         89672                  недвижимость                     C  
21523        244093  на покупку своего автомобиля                     B  
21524         82047         на покупку автомобиля                     C  

[21331 rows x 13 columns]

**Задание 18. Создайте функцию, которая на основании данных из столбца `purpose` сформирует новый столбец `purpose_category`, в который войдут следующие категории:**

- `'операции с автомобилем'`,
- `'операции с недвижимостью'`,
- `'проведение свадьбы'`,
- `'получение образования'`.

**Например, если в столбце `purpose` находится подстрока `'на покупку автомобиля'`, то в столбце `purpose_category` должна появиться строка `'операции с автомобилем'`.**

**Используйте собственную функцию с именем `categorize_purpose()` и метод `apply()`. Изучите данные в столбце `purpose` и определите, какие подстроки помогут вам правильно определить категорию.**


In [54]:
def categorize_purpose(row):
    try:
        if "автом" in row:
            return "операции с автомобилем"
        elif "жил" in row or "недвиж" in row:
            return "операции с недвижимостью"
        elif "свад" in row:
            return "проведение свадьбы"
        elif "образов" in row:
            return "получение образования"
    except:
        return "нет категории"

In [55]:
data["purpose_category"] = data["purpose"].apply(categorize_purpose)

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Отлично!✅:</b> Денис, молодец, отлично выполнил предобработку!

</div>


### Шаг 3. Исследуйте данные и ответьте на вопросы


#### 3.1 Есть ли зависимость между количеством детей и возвратом кредита в срок?


In [56]:
# Оценим количество кредитополучателей в зависимости от количества детей и построим сводную таблицу

data_grouped_by_children = data.groupby("children")
number_of_loaners_by_children = data_grouped_by_children["children"].count()

number_of_loaners_on_child_dict = dict(
    [
        (
            "children",
            number_of_loaners_by_children.index.tolist(),
        ),
        ("reciepent_number", number_of_loaners_by_children.values.tolist()),
    ]
)
pivot_table_return_ratio_on_children = pd.DataFrame(
    number_of_loaners_on_child_dict)

pivot_table_return_ratio_on_children

children  reciepent_number
0         0             14091
1         1              4808
2         2              2052
3         3               330
4         4                41
5         5                 9

In [57]:
# Найдем средний процент невозрвата кредита всрок в зависимости от количества детей и количество людей, не вернувших кредит. Добавим их в сводную таблицу

unsuccessful_debt_return_ratio_depend_children = (
    data_grouped_by_children["debt"].sum(
    ) / data_grouped_by_children["debt"].count()
) * 100

pivot_table_return_ratio_on_children[
    "unsuccessful_debt_return_number"
] = data_grouped_by_children["debt"].sum()
pivot_table_return_ratio_on_children[
    "unsuccessful_debt_return_ratio"
] = unsuccessful_debt_return_ratio_depend_children

pivot_table_return_ratio_on_children

children  reciepent_number  unsuccessful_debt_return_number  \
0         0             14091                             1063   
1         1              4808                              444   
2         2              2052                              194   
3         3               330                               27   
4         4                41                                4   
5         5                 9                                0   

   unsuccessful_debt_return_ratio  
0                        7.543822  
1                        9.234609  
2                        9.454191  
3                        8.181818  
4                        9.756098  
5                        0.000000

<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Отлично!✅:</b> Супер, верно сгруппировал данные для анализа!

</div>


<br/>
<div class="alert alert-block alert-danger">
<h2> Комментарий ревьюера <a class="tocSkip"></h2>

<b>Необходимо исправить❌:</b>
Но будет лучше если представишь данные в одной сводной таблице: категория/всего кредитополучателей/всего должников/доля должников. Так даже анализировать будет легче:)

</div>


<div class="alert alert-success">
<h2> Комментарий ревьюера v2<a class="tocSkip"> </h2>

<b>Отлично!✅:</b> Молодец, отличная табличка получилась!
</div>

**Вывод:**
Сделать вывод о том, как часто кредит всрок возвращают люди с 4 и 5-ю детьми невозможно из-за недостатка данных.
Видим, что заемщики без детей в среднем на 1-2 процента чаще возвращают кредит всрок, что может быть связано с отсутствием расходов на детей.

Upd:
Люди с 1 ребенком или 2 детьми не возвращают кредит практически одинаково часто (9.3% случаев). Лица с 3-мя детьми детьми не возвращают кредит чуть чаще (8.5% случаев).


<div class="alert alert-warning">
    <h2> Комментарий ревьюера <a class="tocSkip"> </h2>
    
<b>Можно лучше⚠️:</b> Ты правильно обратил внимание на то, что выборки несбалансированны и сложно сделать общий вывод. Поэтому лучше, для выявления зависимости, сравнить самые многочисленные категории - семьи с 0, 1 и 2 детьми. У тебя не хватает общего вывода, если мы все обобщим то, что наблюдаем с ростом количества детей в семьях?
</div>


<div class="alert alert-success">
<h2> Комментарий ревьюера v2<a class="tocSkip"> </h2>

<b>Отлично!✅:</b> Можем также наблюдать, что с увеличением количества детей в семьях растут и задолженности!
</div>

#### 3.2 Есть ли зависимость между семейным положением и возвратом кредита в срок?


In [58]:
# Оценим количество заемщиков в зависимости от семейного положения и построим сводную таблицу

data_grouped_by_family_status = data.groupby("family_status")

number_of_loaners_by_family_status = data_grouped_by_family_status[
    "family_status"
].count()

pivot_table_return_ratio_on_family_status = pd.DataFrame(
    columns=["family_statuts", "reciepent_number"]
)
pivot_table_return_ratio_on_family_status[
    "family_statuts"
] = number_of_loaners_by_family_status.index.tolist()
pivot_table_return_ratio_on_family_status[
    "reciepent_number"
] = number_of_loaners_by_family_status.values.tolist()

pivot_table_return_ratio_on_family_status

family_statuts  reciepent_number
0  Не женат / не замужем              2796
1              в разводе              1189
2         вдовец / вдова               951
3       гражданский брак              4134
4        женат / замужем             12261

In [59]:
# Найдем средний процент невозрвата кредита всрок в зависимости от семейного положения и количество людей, не вернувших кредит, добавим их в сводную таблицу

unsuccessful_debt_return_ratio_depend_family_status = (
    data_grouped_by_family_status["debt"].sum() /
    data_grouped_by_family_status["debt"].count()) * 100

pivot_table_return_ratio_on_family_status[
    "unsuccessful_debt_return_number"] = (
        data_grouped_by_family_status["debt"].sum().tolist())
pivot_table_return_ratio_on_family_status[
    "unsuccessful_debt_return_ratio"] = unsuccessful_debt_return_ratio_depend_family_status.tolist(
)
pivot_table_return_ratio_on_family_status = (
    pivot_table_return_ratio_on_family_status.sort_values(
        by="unsuccessful_debt_return_ratio"))
pivot_table_return_ratio_on_family_status.reset_index(drop=True, inplace=True)

pivot_table_return_ratio_on_family_status

family_statuts  reciepent_number  unsuccessful_debt_return_number  \
0         вдовец / вдова               951                               63   
1              в разводе              1189                               84   
2        женат / замужем             12261                              927   
3       гражданский брак              4134                              385   
4  Не женат / не замужем              2796                              273   

   unsuccessful_debt_return_ratio  
0                        6.624606  
1                        7.064760  
2                        7.560558  
3                        9.313014  
4                        9.763948

<br/>
<div class="alert alert-block alert-danger">
<h2> Комментарий ревьюера <a class="tocSkip"></h2>

<b>Необходимо исправить❌:</b>
Денис, также, как и выше, необходимо объеденить данные в общую сводную таблицу. Откорректируй, пожалуйста, данный недочет во всей работе.

</div>


<div class="alert alert-success">
<h2> Комментарий ревьюера v2<a class="tocSkip"> </h2>

<b>Отлично!✅:</b> +
</div>

**Вывод:**
Видим, что люди в статусе вдовца / вдовы наиболее успешно возвращают кредиты, что может быть связано с полученным наследством или с тем, что они брали кредит на похороны. Часто, такие кредиты компесируются родственниками
Люди в в браке и в разводе возвращают кредиты на 2% лучше чем люди не замужем или в гражданском браке, и на 1% хуже чем люди потерявшие супругу / супруга


<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Отлично!✅:</b> Абсолютно правильно, заемщики состоящие или состоявшие в браке реже допускают просрочку!

</div>


#### 3.3 Есть ли зависимость между уровнем дохода и возвратом кредита в срок?


In [60]:
# Оценим количество заемщиков в зависимости от дохода и построим сводную таблицу

data_grouped_by_income = data.groupby("total_income_category")
number_of_loaners_by_income = data_grouped_by_income["total_income_category"].count(
)

pivot_table_return_ratio_on_income = pd.DataFrame(
    columns=["income_category", "income_amount", "reciepent_number"]
)
pivot_table_return_ratio_on_income[
    "income_category"
] = number_of_loaners_by_income.index.tolist()
pivot_table_return_ratio_on_income["income_amount"] = [
    "1000001 и выше",
    "200001–1000000",
    "50001–200000",
    "30001–50000",
    "0–30000",
]
pivot_table_return_ratio_on_income[
    "reciepent_number"
] = number_of_loaners_by_income.values.tolist()

pivot_table_return_ratio_on_income

income_category   income_amount  reciepent_number
0               A  1000001 и выше                25
1               B  200001–1000000              5014
2               C    50001–200000             15921
3               D     30001–50000               349
4               E         0–30000                22

In [61]:
# Найдем средний процент возрвата кредита всрок в зависимости от дохода и количество людей не вернувших кредит, добавим их в сводную таблицу

unsuccessful_debt_return_ratio_depend_income = (
    data_grouped_by_income["debt"].sum(
    ) / data_grouped_by_income["debt"].count()
) * 100

pivot_table_return_ratio_on_income["unsuccessful_debt_return_number"] = (
    data_grouped_by_income["debt"].sum().tolist()
)
pivot_table_return_ratio_on_income[
    "unsuccessful_debt_return_ratio"
] = unsuccessful_debt_return_ratio_depend_income.tolist()
pivot_table_return_ratio_on_income = pivot_table_return_ratio_on_income.sort_values(
    by="unsuccessful_debt_return_ratio"
)
pivot_table_return_ratio_on_income.reset_index(drop=True, inplace=True)

pivot_table_return_ratio_on_income

income_category   income_amount  reciepent_number  \
0               D     30001–50000               349   
1               B  200001–1000000              5014   
2               A  1000001 и выше                25   
3               C    50001–200000             15921   
4               E         0–30000                22   

   unsuccessful_debt_return_number  unsuccessful_debt_return_ratio  
0                               21                        6.017192  
1                              354                        7.060231  
2                                2                        8.000000  
3                             1353                        8.498210  
4                                2                        9.090909

In [62]:
# Данных по категориям A, E недостаточно, чтобы сделать выводы, поэтому удалим их из сводной таблицы

pivot_table_return_ratio_on_income = pivot_table_return_ratio_on_income.loc[
    (pivot_table_return_ratio_on_income["income_category"] != "A")
    & (pivot_table_return_ratio_on_income["income_category"] != "E")
]

pivot_table_return_ratio_on_income

income_category   income_amount  reciepent_number  \
0               D     30001–50000               349   
1               B  200001–1000000              5014   
3               C    50001–200000             15921   

   unsuccessful_debt_return_number  unsuccessful_debt_return_ratio  
0                               21                        6.017192  
1                              354                        7.060231  
3                             1353                        8.498210

**Вывод:**
Данных по категории А и Е слишком мало для точных выводов. Наиболее успешно кредиты возвращают люди с доходом 30001–50000, что может быть связано с маленькими сумамаи кредита (случаев невозврата 6%)

Upd:

Для категорий B и C можем заметить, что с ростом дохода уменьшается количество возвратов всрок (Для D 91.5%, а для B 93%)


<div class="alert alert-warning">
    <h2> Комментарий ревьюера <a class="tocSkip"> </h2>
    
<b>Можно лучше⚠️:</b> Давай сравним самые крупные выборки В и С для выявления зависимости.
</div>


<div class="alert alert-success">
<h2> Комментарий ревьюера v2<a class="tocSkip"> </h2>

<b>Отлично!✅:</b> Ты выбрал три категории для сравнения, D очень маленькая категория, относительно остальных. Поэтому лучше сравнить самые многочисленные категории - В и С, тогда увидим, что чем выше доход, тем надежнее заемщик!
</div>

#### 3.4 Как разные цели кредита влияют на его возврат в срок?


In [63]:
# Оценим количество заемщиков в зависимости от цели кредита и построим сводную таблицу

data_grouped_by_purpose_category = data.groupby("purpose_category")
data_grouped_by_purpose_category["purpose_category"].count()
number_of_loaners_by_purpose_category = data_grouped_by_purpose_category[
    "purpose_category"].count()

pivot_table_return_ratio_on_purpose_category = pd.DataFrame(
    columns=["purpose_category", "reciepent_number"])
pivot_table_return_ratio_on_purpose_category[
    "purpose_category"] = number_of_loaners_by_purpose_category.index.tolist()

pivot_table_return_ratio_on_purpose_category[
    "reciepent_number"] = number_of_loaners_by_purpose_category.values.tolist(
)

pivot_table_return_ratio_on_purpose_category

purpose_category  reciepent_number
0    операции с автомобилем              4279
1  операции с недвижимостью             10751
2     получение образования              3988
3        проведение свадьбы              2313

In [64]:
# Найдем средний процент не возврата кредита всрок в зависимости от цели кредита и количество людей не вернувших кредит, добавим их в сводную таблицу


unsuccessful_debt_return_ratio_depend_purpose_category = (
    data_grouped_by_purpose_category["debt"].sum()
    / data_grouped_by_purpose_category["debt"].count()
    * 100
)

pivot_table_return_ratio_on_purpose_category["unsuccessful_debt_return_number"] = (
    data_grouped_by_purpose_category["debt"].sum().tolist()
)
pivot_table_return_ratio_on_purpose_category[
    "unsuccessful_debt_return_ratio"
] = unsuccessful_debt_return_ratio_depend_purpose_category.tolist()
pivot_table_return_ratio_on_purpose_category = (
    pivot_table_return_ratio_on_purpose_category.sort_values(
        by="unsuccessful_debt_return_ratio"
    )
)
pivot_table_return_ratio_on_purpose_category.reset_index(
    drop=True, inplace=True)

pivot_table_return_ratio_on_purpose_category

purpose_category  reciepent_number  \
0  операции с недвижимостью             10751   
1        проведение свадьбы              2313   
2     получение образования              3988   
3    операции с автомобилем              4279   

   unsuccessful_debt_return_number  unsuccessful_debt_return_ratio  
0                              780                        7.255139  
1                              183                        7.911803  
2                              369                        9.252758  
3                              400                        9.347978

**Вывод:**
Кредиты, взятые на недвижимость, возвращают чаще всего, на 0.7 % реже возвращают кредиты на свадьбу. Кредиты на автомобиль и образование возвращают на 2% реже чем кредиты на недвижимость.


<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Отлично!✅:</b> Молодец, верно определил более и менее рискованные для банка категории заемщиков по целям кредита!

</div>


#### 3.5 Приведите возможные причины появления пропусков в исходных данных.


_Ответ:_ Заметим, что пропуски данных о доходе и месте работы появляются вместе

Это может быть по тому, что:

1. Люди могут не хотеть раскрывать свое место работы и свой доход
2. Люди не могут подвердить свое место работы и доход
3. Произошла техническая ошибка


<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Отлично!✅:</b> Совершенно верно, на пропуски в данных, в основном, оказывают влияние человеческий и технический факторы!

</div>


#### 3.6 Объясните, почему заполнить пропуски медианным значением — лучшее решение для количественных переменных.


_Ответ:_
Потому что очень часто в количественных переменных встречаются аномально большие или маленькие значения. Например, у 1% самых богатых людей на планете средств больше чем у остального населения планеты, если взять среднее состояние одного человека от этих данных мы получим неверные значения


<div class="alert alert-success">
<h2> Комментарий ревьюера <a class="tocSkip"> </h2>

<b>Отлично!✅:</b> Если будем точнее, то медиана менее чувствительна к выбросам!

</div>


## Шаг 4: общий вывод.


### Данные для анализа и их предобработка

Анализ был проведен на данных о 21525 заемщиков кредитов. Был удален 71 дупликат.

##### Были решены следующие проблемы во входных данных:

1. Отрицательные значения в столбце days_employed сделаны положительными.
2. Для каждого из типов дохода пропуски в столбцах total_income и days_emloyed были заполнены медианными значениями этой группы
3. Удалены строки с аномальным количеством детей (20, -1)

##### Изменения в данных

1. Добавлен стоблец с категорией уровня дохода total_income_category и столбец с категорией цели кредита purpose_category

##### Рекомендации по сбору данных

1. Наладить автоматизированную систему так, чтобы в данных не возникало отрицательных чисел
2. Добавить в опрос пользователя обязательный пункт "Цель кредита"
3. Записывать все качественный описания пользователя в нижнем регистре (образование не Высшее, а высшее)

##### 3.5 Возможные причины появления пропусков в исходных данных.

Пропуски есть лишь в столбцах days_employed и total_income

Это может быть по тому, что:

1. Люди могут не хотеть раскрывать свое место работы и свой доход
2. Люди не могут подвердить свое место работы и доход
3. Произошла техническая ошибка

### Итоги анализа

##### Есть ли зависимость между количеством детей и возвратом кредита в срок?

Сделать вывод о том, как часто кредит всрок возвращают люди с 4 и 5-ю детьми невозможно из-за недостатка данных.
Заемщики без детей в среднем на 1-2 процента чаще возвращают кредит всрок, что может быть связано с отсутствием расходов на детей.
Люди с 1 ребенком или 2 детьми не возвращают кредит практически одинаково часто (9.3% случаев). Лица с 3-мя детьми детьми не возвращают кредит чуть чаще (8.5% случаев).

##### 3.2 Есть ли зависимость между семейным положением и возвратом кредита в срок?

Люди в статусе вдовца / вдовы наиболее успешно возвращают кредиты, что может быть связано с полученным наследством или с тем, что они брали кредит на похороны. Часто, такие кредиты компесируются родственниками. _(невозврат в 6.6% случаев)_
Люди в в браке _(невозврат в 7.6% случаев)_ и в разводе _(невозврат в 7% случаев)_ возвращают кредиты на 2% лучше чем люди не замужем _(невозврат в 9.7% случаев)_ или в гражданском браке _(невозврат в 9.5% случаев)_, и на 1% хуже чем люди потерявшие супругу / супруга _(невозврат в 6.6% случаев)_.

##### 3.3 Есть ли зависимость между уровнем дохода и возвратом кредита в срок?

Данных по категории А и Е слишком мало для точных выводов. Наиболее успешно кредиты возвращают люди с доходом 30001–50000, что может быть связано с маленькими сумамаи кредита _(случаев невозврата 6%)_
Для категорий B и C можем заметить, что с ростом дохода уменьшается количество возвратов всрок _(Для D 91.5%, а для B 93%)_

##### 3.4 Как разные цели кредита влияют на его возврат в срок?

Кредиты, взятые на недвижимость, возвращают чаще всего _(невозврат в 7.2% случаях)_, на 0.7 % реже возвращают кредиты на свадьбу. Кредиты на автомобиль и образование возвращают на 2% реже чем кредиты на недвижимость _(невозврат в 9.3% случаях)_.


<br/>
<div class="alert alert-block alert-danger">
<h2> Комментарий ревьюера <a class="tocSkip"></h2>
<b>Необходимо исправить❌:</b>
Денис, общий вывод - обязательный и важный элемент работы. Его нужно сформулировать, взяв основные результаты из всех шагов проекта. Начать нужно с того, с какими данными мы работали и какие действия провели на этапе предобработки, а закончить выявленными при ответе на вопросы зависимостями и рекомендациями заказчику по улучшению сбора данных и скоринга. Заказчик, зачастую, читает только общий вывод исследования - поэтому он должен быть максимально информативным. 
</div>


<div class="alert alert-success">
<h2> Комментарий ревьюера v2<a class="tocSkip"> </h2>

<b>Отлично!✅:</b> Денис, привет! Молодец, верно доработал проект. Работа принята, хорошего дня:)
</div>

<div style="border:solid Chocolate 2px; padding: 40px">
Денис, давай подведем итоги. Ты большой молодец, что старался и проделал такую большую работу. Шаг 1 и 2 ты выполнил без проблем, что очень хорошо:)
    
    Разберем над чем стоит поработать:
    
    1. Необходимо добавить сводные таблицы в проект для анализа. Для этого можно использовать метод groupby() или pivot_table().
    
    2. Оформить общий вывод.
    
    3. Рекомендую:
    - учитывать несбалансированность выборок.
    
    Пересмотри работу и исправь ошибки. Обязательно пиши, если я могу тебе еще чем-то помочь. Удачи!)
